mlx-rs/src/array/mod.rs

```diff
@@ -439,29 +412,46 @@ impl Array {
                 }
             };
 
-            let new_c_array = mlx_sys::mlx_array_from_data(
-                data,
-                shape.as_ptr(),
-                shape.len() as i32,
-                dtype.into(),
-            );
+            let new_c_array =
+                mlx_sys::mlx_array_new_data(data, shape.as_ptr(), shape.len() as i32, dtype.into());
 
             Array::from_ptr(new_c_array)
         }
     }
 }
 
+impl Clone for Array {
+    fn clone(&self) -> Self {
+        unsafe {
+            let mut c_array = mlx_sys::mlx_array_new();
+            // `mlx_array_set` should invoke the copy constructor of the array,
+            // which contains a shared_ptr to the data.
+            mlx_sys::mlx_array_set(&mut c_array as *mut _, self.c_array);
+            Array::from_ptr(c_array)
```

Comment:
If you look into the internal of `mlx_array_set`, it would only throw if, in this case, `self.c_array` is null, and the user would not be able to construct an `Array` with a `null` without using `unsafe`. But I guess we could do a `debug_assert` here